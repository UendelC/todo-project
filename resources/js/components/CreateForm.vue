<template>
    <div class="form__container">
        <create-input
            v-model="title"
            id="title"
            label="Título"
        />
        <create-input
            v-model="description"
            id="description"
            label="Descrição"
        />
        <div class="form__button">
            <cds-button
                variant="blue"
                size="lg"
                text="Adicionar novo"
                :disabled="disableButton"
                @click="handleSubmit"
            />
        </div>
    </div>
</template>

<script>
import CreateInput from './CreateInput.vue';

export default {
    components: {
        CreateInput,
    },

    data() {
        return {
            title: '',
            description: '',
        };
    },

    computed: {
        disableButton() {
            return this.title === '' || this.description === '';
        },
    },

    methods: {
        handleSubmit() {
            if (this.title !== '' && this.description !== '') {
                const emitData = {
                    title: this.title,
                    description: this.description,
                };
                this.$emit('submit', emitData);
            }
        },
    },
}
</script>

<style lang="scss" scoped>
@import 'node_modules/@sysvale/cuida/dist/@sysvale/tokens.scss';

.form {
    &__container {
        width: 50vw;
    }

    &__button {
        margin: my(6);
    }
}
</style>
