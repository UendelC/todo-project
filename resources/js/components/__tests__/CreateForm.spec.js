import { test, describe, expect } from "vitest";
import CreateForm from '../CreateForm.vue';
import { shallowMount, flushPromises, mount } from "@vue/test-utils";

describe('CreateForm', () => {
    test('component renders correctly', async () => {
        const wrapper = shallowMount(CreateForm, () => {});

        await flushPromises();

        expect(wrapper).toMatchSnapshot();
    });

    test('submit button is enabled when the form is filled', async () => {
        expect.assertions(2);

        const wrapper = shallowMount(CreateForm, () => {});

        await flushPromises();

        expect(wrapper.vm.disableButton).toBeTruthy();

        await wrapper.setData({
            title: 'Test title',
            description: 'Test description',
        });

        expect(wrapper.vm.disableButton).toBeFalsy();
    });

    test('data is emitted correctly on submit', async () => {
        const wrapper = mount(CreateForm, {});

        wrapper.setData({
            title: 'Test title',
            description: 'Test description',
        });

        wrapper.vm.handleSubmit();

        expect(wrapper.emitted()).toHaveProperty('submit');
        expect(wrapper.emitted('submit')[0]).toContainEqual({
            title: 'Test title',
            description: 'Test description',
        });
    });
})
