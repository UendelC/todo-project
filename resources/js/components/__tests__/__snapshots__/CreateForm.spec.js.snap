// Vitest Snapshot v1

exports[`CreateForm > component renders correctly 1`] = `
VueWrapper {
  "__app": {
    "_component": {
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <div
        class="form__container"
      >
        <create-input … />
        <create-input … />
        <div … />
      </div>
    </div>,
    "_context": {
      "app": [Circular],
      "components": {
        "transition": [Object],
        "transition-group": [Object],
      },
      "config": {
        "compilerOptions": [Object],
        "errorHandler": undefined,
        "globalProperties": [Object],
        "isNativeTag": [Function],
        "optionMergeStrategies": [Object],
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": {},
      "emitsCache": WeakMap {},
      "mixins": [
        [Object],
      ],
      "optionsCache": WeakMap {},
      "propsCache": WeakMap {},
      "provides": {},
      "reload": [Function],
    },
    "_instance": {
      "a": null,
      "accessCache": {},
      "appContext": {
        "app": [Circular],
        "components": [Object],
        "config": [Object],
        "directives": [Object],
        "emitsCache": WeakMap {},
        "mixins": [Array],
        "optionsCache": WeakMap {},
        "propsCache": WeakMap {},
        "provides": [Object],
        "reload": [Function],
      },
      "asyncDep": null,
      "asyncResolved": false,
      "attrs": {},
      "bc": null,
      "bm": null,
      "bu": null,
      "bum": null,
      "c": null,
      "components": null,
      "ctx": {},
      "da": null,
      "data": {},
      "directives": null,
      "ec": null,
      "effect": ReactiveEffect {
        "active": true,
        "allowRecurse": true,
        "deps": [Array],
        "fn": [Function],
        "onTrack": undefined,
        "onTrigger": undefined,
        "parent": undefined,
        "scheduler": [Function],
      },
      "emit": [Function],
      "emitsOptions": null,
      "emitted": null,
      "exposeProxy": null,
      "exposed": null,
      "inheritAttrs": undefined,
      "isDeactivated": false,
      "isMounted": true,
      "isUnmounted": false,
      "m": null,
      "next": null,
      "parent": null,
      "props": {},
      "propsDefaults": {},
      "propsOptions": [
        [Object],
        [Array],
      ],
      "provides": {},
      "proxy": {},
      "refs": {
        "VTU_COMPONENT": [Object],
      },
      "render": [Function],
      "renderCache": [],
      "root": [Circular],
      "rtc": null,
      "rtg": null,
      "scope": EffectScope {
        "active": true,
        "cleanups": [Array],
        "effects": [Array],
      },
      "setupContext": null,
      "setupState": {},
      "slots": {},
      "sp": null,
      "subTree": {
        "__v_isVNode": true,
        "__v_skip": true,
        "anchor": null,
        "appContext": null,
        "children": null,
        "component": [Object],
        "dirs": null,
        "dynamicChildren": null,
        "dynamicProps": null,
        "el": <div … />,
        "key": null,
        "patchFlag": 0,
        "props": [Object],
        "ref": [Object],
        "scopeId": null,
        "shapeFlag": 4,
        "slotScopeIds": null,
        "ssContent": null,
        "ssFallback": null,
        "staticCount": 0,
        "suspense": null,
        "target": null,
        "targetAnchor": null,
        "transition": null,
        "type": [Object],
      },
      "suspense": null,
      "suspenseId": 0,
      "type": {
        "name": "VTU_ROOT",
        "render": [Function],
      },
      "u": null,
      "uid": 0,
      "um": null,
      "update": [Function],
      "vnode": {
        "__v_isVNode": true,
        "__v_skip": true,
        "anchor": null,
        "appContext": [Object],
        "children": null,
        "component": [Circular],
        "dirs": null,
        "dynamicChildren": null,
        "dynamicProps": null,
        "el": <div … />,
        "key": null,
        "patchFlag": 0,
        "props": null,
        "ref": null,
        "scopeId": null,
        "shapeFlag": 4,
        "slotScopeIds": null,
        "ssContent": null,
        "ssFallback": null,
        "staticCount": 0,
        "suspense": null,
        "target": null,
        "targetAnchor": null,
        "transition": null,
        "type": [Object],
      },
      "withProxy": null,
    },
    "_props": null,
    "_uid": 0,
    "component": [Function],
    "config": {
      "compilerOptions": {},
      "errorHandler": undefined,
      "globalProperties": {},
      "isNativeTag": [Function],
      "optionMergeStrategies": {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.2.37",
  },
  "__setProps": [Function],
  "componentVM": {
    "description": "",
    "disableButton": true,
    "handleSubmit": [Function],
    "hasOwnProperty": [Function],
    "title": "",
  },
  "isDisabled": [Function],
  "rootVM": {},
  "wrapperElement": <div
    class="form__container"
  >
    <create-input
      id="title"
      label="Título"
      modelvalue=""
    />
    <create-input
      id="description"
      label="Descrição"
      modelvalue=""
    />
    <div
      class="form__button"
    >
      <cds-button
        disabled="true"
        size="lg"
        text="Adicionar novo"
        variant="blue"
      />
    </div>
  </div>,
}
`;
